### PR TITLE
implement local event handling

### PR DESCRIPTION
**Pull request by imcaizheng**
## Notes
- The event handling is implemented in an asynchronous way so that it conforms to the original requirement mentioned on https://github.com/topcoder-platform/taas-apis/issues/54, and, the logic is still decoupled in the meantime.
- The logic at `src/app-routes.js` and inside function `ResourceBookingService.updateResourceBooking` are slightly altered
    to work with the event handling. Please review these changes particularly to ensure nothing would be broken due to that.

### CODE REVIEW

**Review by maxceem — CHANGES_REQUESTED** [review]
@imcaizheng, automatic updates doesn't work for me. When I make any action which supposes to trigger automatic updates, I get error.

1. When I update ResourceBooking status to `assigned` I get an error:

    image

    - To reproduce I create Job, JobCandidate and ResourceBooking using the same userId, to trigger automatic status update logic.

2. If I change the status of Job to `cancelled` I get an error:

     image

File: app-routes.js [hunk 1]
```diff
@@ -43,8 +43,7 @@ module.exports = (app) => {
             if (!req.authUser.scopes || !helper.checkIfExists(def.scopes, req.authUser.scopes)) {
               next(new errors.ForbiddenError('You are not allowed to perform this action!'))
             } else {
-              req.authUser.userId = config.m2m.M2M_AUDIT_USER_ID
-              req.authUser.handle = config.m2m.M2M_AUDIT_HANDLE
+              req.authUser = helper.authUserAsM2M()
```
[the remark this answers is not in the document]
Comment (imcaizheng) [reply app-routes.js]: @maxceem PR updated as per your suggestions.

**Comment by imcaizheng** [comment]
@maxceem I tried to resolve the conflict between the PR and dev branch but I did it wrongly. Please discard the last commit and merge the last but one, named `revert the change made to app-routes.js`.

**Comment by maxceem** [comment]
@imcaizheng hmm, I don't think there is a way to discard the last commit during merging. Could you please, revert the wrong commit from `imcaizheng:implement-event-handlers` branch, so it's not included in the PR?

**Comment by imcaizheng** [comment]
@maxceem Thanks. The change is reverted.

**Review by maxceem — CHANGES_REQUESTED** [review]
> @maxceem Thanks. The change is reverted.

@imcaizheng Looks like during merging changes that were done for permissions in this PR https://github.com/topcoder-platform/taas-apis/pull/77 were reverted.

image

I try to test functionality anyway:
- When I assign Resouce Booking the corresponding JobCandidate became 'selected'.
- But if I cancel Job, corresponding Resource Booking and JobCandidates are not cancelled.

**Comment by imcaizheng** [comment]
@maxceem
Please accept this PR https://github.com/topcoder-platform/taas-es-processor/pull/5 for `taas-es-processor` and try again. Resource Booking and JobCandidates could be properly cancelled after that.

**Pull request closed without merging**